marlowe/specification/ReadMe.md

```diff
@@ -0,0 +1,346 @@
+  where isMarloweAddress (Address (ScriptCredential hash)) = hash == marloweValidatorHash
+        isMarloweAddress _ = False
+```
+
+
+#### *Constraint 5.* Input value from script
+
+The value of the Marlowe UTxO being spent exactly matches the value recorded in the Marlowe state.
+```haskell
+inValue ≡ foldMap toValue (accounts $ marloweState marloweData)
+  where toValue ((_, Token currency name), count)) = Value.singleton currency name count
+```
+
+
+#### *Constraint 6.* Output value to script
+
+The beginning balance plus the deposits equals the ending balance plus the payments.
+```haskell
+inValue + foldMap valueOfDeposit (fmap getInputContent transactionInput) ≡ outValue + foldMap valueOfPayment (txOutPayments transactionOutput)
+  where getInputContent (Input inputContent) = inputContent
+        getInputContent (MerkleizedTxInput inputContent _) = inputContent
+        valueOfDeposit (IDeposit _ _ (Token currency name) count) = Value.singleton currency name count
+        valueOfDeposit _ = mempty
+        valueOfPayment (Payment _ (Party _) value)) = value
+        valueOfPayment _ = mempty
+```
+
+
+#### *Constraint 7.* Input state
+
+The input state of the transaction is used in the semantics computation.
+```haskell
+marloweState marloweData ≡ inState
+```
+
+
+#### *Constraint 8.* Input contract
+
+The input contract of the transaction is used in the semantics computation.
+```haskell
+marloweContract marloweData ≡ inContract
+```
+
+
+#### *Constraint 9.* Marlowe parameters
+
+The Marlowe parameters are not changed by the transaction.
+```haskell
+Just $ marloweParams marloweData ≡ fmap marloweParams . fromBuiltinData $ getDatum outDatum
+```
+
+
+#### *Constraint 10.* Output state
+
+The output state of the transaction matches the result of the semantics computation.
+```haskell
+Just $ txOutState transactionOutput ≡ fmap marloweState . fromBuiltinData $ getDatum outDatum
+```
+
+
+#### *Constraint 11.* Output contract
+
+The output contract of the transaction matches the result of the semantics computation.
+```haskell
+Just $ txOutContract transactionOutput ≡ fmap marloweContract . fromBuiltinData $ getDatum outDatum
+```
+
+
+#### *Constraint 12.* Merkleized continuations
+
+All of the merkleized contracts referenced in the input are provided to the script.
+```haskell
+all demerkleizes transactionInput ≡ True
+  where demerkleizes (MerkelizedTxInput _ hash) = isJust (fromBuiltinData =<< lookup hash (txInfoData $ scriptContextTxInfo scriptContext) :: Maybe Contract)
+        demerkleizes (Input _) = True
+```
+
+
+#### *Constraint 13.* Positive balances
+
+All accounts in the state have positive balances.
+```haskell
+all ((> 0) . snd) (accounts $ marloweState marloweData) ≡ True
+```
+
+
+#### *Constraint 14.* Inputs authorized
+
+The required signature or at-least-one role token is present in the transaction for each deposit or choice.
```

Comment:
why hyphenated at-least-one here? Is it a technical term?

Reply:
fixed in 8be04323e4ec24c4e923ed214094e85c916ab8e0